File: Algorithms/C/16 3Sum Closest.c
int compare (const void * a, const void * b){
  return ( *(int*)a - *(int*)b );
}

int threeSumClosest(int* nums, int numsSize, int target){
    int ans = nums[0] + nums[1] + nums[2];
    qsort(nums, numsSize, sizeof(int), compare);
    int s;
    for (int i = 0; i < numsSize - 2; ++i) {
        int j, k;
        j = i + 1, k = numsSize - 1;
        while (j < k) {
            s = nums[i] + nums[j] + nums[k];
            if (abs(s - target) < abs(ans - target)) {
                ans = s;
            }
            if (s == target) {
                return target;
            }
            else if (s < target) {
                ++j;
            }
            else {
                --k;
            }
        }
    }
    return ans;
}

File: Algorithms/Python/16 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        ans = nums[0] + nums[1] + nums[2]
        n = len(nums)
        for i in range(n - 2):
            j, k = i + 1, n - 1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                if abs(s - target) < abs(ans - target):
                    ans = s
                if s == target:
                    return target
                elif s < target:
                    j += 1
                else:
                    k -= 1
        return ans
